Askrene more tweaks

rustyrussell opened the pull request:
OK, lots of real-life testing for askrene, resulting in a number of bug fixes, including a nasty cycle fix from @Lagrang3  Looking forward to @Lagrang3 review!

(I need to benchmark again, too!)

Lagrang3 commented on doc/schemas/lightning-getroutes.json:
```diff
@@ -14,7 +14,7 @@
     "Layers are generally maintained by plugins, either to contain persistent information about capacities which have been discovered, or to contain transient information for this particular payment (such as blinded paths or routehints).",
     "",
-    "There are two automatic layers: *auto.localchans* contains information on local channels from this node (including non-public ones), and their exact current spendable capacities, and *auto.sourcefree* overrides all channels leading out of the *source* to be zero fee and zero delay.  These are both useful in the case where the source is the current node."
+    "There are two automatic layers: *auto.localchans* contains information on local channels from this node (including non-public ones), and their exact current spendable capacities, and *auto.sourcefree* overrides all channels (including those from prevous layers) leading out of the *source* to be zero fee and zero delay.  These are both useful in the case where the source is the current node."
```
```suggestion
    "There are two automatic layers: *auto.localchans* contains information on local channels from this node (including non-public ones), and their exact current spendable capacities, and *auto.sourcefree* overrides all channels (including those from previous layers) leading out of the *source* to be zero fee and zero delay.  These are both useful in the case where the source is the current node."
```

Lagrang3 commented on plugins/askrene/askrene.c:
```diff
@@ -409,10 +535,17 @@ static const char *get_routes(const tal_t *ctx,
 			rh->delay = delay;
 		}
 		(*amounts)[i] = flows[i]->delivers;
+		struct amount_msat fee;
+		if (!amount_msat_sub(&fee, r->hops[0].amount, flows[i]->delivers))
+			abort();
+		rq_log(tmpctx, rq, LOG_INFORM, "Flow %zu/%zu: %s",
+		       i, tal_count(flows),
+		       fmt_route(tmpctx, r, (*amounts)[i], finalcltv));
```
`fee` is not used here, maybe you wanted to show fees in the logs?

rustyrussell replied:
Oops, part of my debugging before I decided to log the whole thing!  Removed, thanks!

Lagrang3 commented on plugins/askrene/explain_failure.c:
```diff
@@ -236,7 +236,8 @@ const char *explain_failure(const tal_t *ctx,
 				   AMOUNT_MSAT(0), 0);
 	if (!hops)
-		return tal_fmt(ctx, "There is no connection between source and destination at all");
+		return rq_log(ctx, rq, LOG_INFORM,
```
I am not sure we should send logs from inside `explain_failure`. Maybe we should only use it
to get useful messages. Then the caller will make the logs.

On the other hand, if we are going to let `explain_failure` to write logs we might as well need to log the failure branches in lines 223 and 229.

rustyrussell replied:
Yes, I started doing this when I was trying to reconstruct what happens from the debug logs: too much was missing.  It probably does need to go deeper, you're right!

rustyrussell commented:
I wonder about using "double" for network costs instead of s64?  I had an issue with giant `k`, causing an overflow in linearize_channel:

It doesn't seem to be an issue with fixed k, though.

```
		cost[i] = params->cost_fraction[i]
		          *params->amount.millisatoshis /* Raw: linearize_channel */
		          *params->k_factor/(b-a);
```

rustyrussell commented:
Rebase with minor changes (and whitespace fix which was stopping CI).

Lagrang3 commented:
> I wonder about using "double" for network costs instead of s64? I had an issue with giant `k`, causing an overflow in linearize_channel:
> 
> It doesn't seem to be an issue with fixed k, though.
> 
> ```
> 		cost[i] = params->cost_fraction[i]
> 		          *params->amount.millisatoshis /* Raw: linearize_channel */
> 		          *params->k_factor/(b-a);
> ```

The theory bounds the complexity of the MCF algorithms based on the assumption that costs, capacities and demand/supply are all integers.

rustyrussell commented:
> > I wonder about using "double" for network costs instead of s64? I had an issue with giant `k`, causing an overflow in linearize_channel:
> > It doesn't seem to be an issue with fixed k, though.
> > ```
> > 		cost[i] = params->cost_fraction[i]
> > 		          *params->amount.millisatoshis /* Raw: linearize_channel */
> > 		          *params->k_factor/(b-a);
> > ```
> 
> The theory bounds the complexity of the MCF algorithms based on the assumption that costs, capacities and demand/supply are all integers.

Hmm, true.  I need to think harder about the constraints here if we want to vary k again, though.  It's multiplied by the product of two amounts, though one is scaled down to sats: that makes me a little nervous!

Lagrang3 commented:
No worries. I am working on a new MCF solver. We will have more control on the base unit.

Lagrang3 commented on plugins/askrene/askrene.c:
```diff
@@ -275,7 +361,8 @@ static const char *get_routes(const tal_t *ctx,
 				/* Handled below, after other layers */
 				assert(streq(layers[i], "auto.sourcefree"));
-				continue;
+				plugin_log(rq->plugin, LOG_DBG, "Adding auto.sourcefree");
+				l = source_free_layer(layers, askrene->gossmap, source, localmods);
```
So the previous version would first update localmods and after the loop it would apply `auto.sourcefree` if present.
Now we have to take care about the order of the layers again, because if `auto.localchans` comes after `auto.sourcefree` in the list then we will not be able to nullify the fees on private local channels, they don't "exist" yet in localmods.

Lagrang3 replied:
I didn't see the comment in the commit: "`getroutes` now applies `auto.sourcefree` layer in the order specified, so doesn't alter channels changed in later layers."

rustyrussell commented:
Added a new method for deriving k which should be more stable (and indeed, other changes kept breaking tests which didn't use the real data, as the k value of 8 was too different).  Also fixed another unrelated bug in reservations.
